### ISSUE TITLE
Hack to find the libsnark include dir is fragile (was build error on xubuntu: "configure: error: libsnark include directory not found")

### ISSUE THREAD

**Issue by zcashuser** [zcash/zcash#1283]
```
~/zcash$ ./zcutil/build.sh -j4
+++ readlink -f ./zcutil/build.sh
++ dirname /home/user/zcash/zcutil/build.sh
+ cd /home/user/zcash/zcutil/..
+ LCOV_ARG=
+ '[' x-j4 = x--enable-lcov ']'
++ pwd
+ PREFIX=/home/user/zcash/depends/x86_64-unknown-linux-gnu/
+ make -j4 -C ./depends/ V=1 NO_QT=1
make: Entering directory `/home/user/zcash/depends'
rm -rf /home/user/zcash/depends/i686-pc-linux-gnu
mkdir -p /home/user/zcash/depends/i686-pc-linux-gnu
echo copying packages: native_ccache boost openssl libsnark libgmp libsodium googletest googlemock bdb miniupnpc
copying packages: native_ccache boost openssl libsnark libgmp libsodium googletest googlemock bdb miniupnpc
echo to: /home/user/zcash/depends/i686-pc-linux-gnu
to: /home/user/zcash/depends/i686-pc-linux-gnu
cd /home/user/zcash/depends/i686-pc-linux-gnu;  tar xf /home/user/zcash/depends/built/i686-pc-linux-gnu/native_ccache/native_ccache-3.1.9-deaa1ee8c66.tar.gz;   tar xf /home/user/zcash/depends/built/i686-pc-linux-gnu/boost/boost-1_57_0-0a7fbf671e7.tar.gz;   tar xf /home/user/zcash/depends/built/i686-pc-linux-gnu/openssl/openssl-1.0.1k-2445bf7e363.tar.gz;   tar xf /home/user/zcash/depends/built/i686-pc-linux-gnu/libsnark/libsnark-0.1-f251f41e259.tar.gz;   tar xf /home/user/zcash/depends/built/i686-pc-linux-gnu/libgmp/libgmp-6.0.0a-80c0a86e8eb.tar.gz;   tar xf /home/user/zcash/depends/built/i686-pc-linux-gnu/libsodium/libsodium-1.0.8-3241aed3009.tar.gz;   tar xf /home/user/zcash/depends/built/i686-pc-linux-gnu/googletest/googletest-1.7.0-b7926635927.tar.gz;   tar xf /home/user/zcash/depends/built/i686-pc-linux-gnu/googlemock/googlemock-1.7.0-50682979e17.tar.gz;   tar xf /home/user/zcash/depends/built/i686-pc-linux-gnu/bdb/bdb-4.8.30-628d55a8e12.tar.gz;   tar xf /home/user/zcash/depends/built/i686-pc-linux-gnu/miniupnpc/miniupnpc-1.9.20151008-a9a6b99ebb9.tar.gz; 
touch /home/user/zcash/depends/i686-pc-linux-gnu/.stamp_746b0b4490e
sed -e 's|@HOST@|i686-pc-linux-gnu|' \
            -e 's|@CC@|gcc -m32|' \
            -e 's|@CXX@|g++ -m32|' \
            -e 's|@AR@|ar|' \
            -e 's|@RANLIB@|ranlib|' \
            -e 's|@NM@|nm|' \
            -e 's|@STRIP@|strip|' \
            -e 's|@build_os@|linux|' \
            -e 's|@host_os@|linux|' \
            -e 's|@CFLAGS@|-pipe -O2|' \
            -e 's|@CXXFLAGS@|-pipe -O2|' \
            -e 's|@CPPFLAGS@||' \
            -e 's|@LDFLAGS@||' \
            -e 's|@no_qt@|1|' \
            -e 's|@no_wallet@||' \
            -e 's|@no_upnp@||' \
            -e 's|@debug@||' \
            config.site.in > /home/user/zcash/depends/i686-pc-linux-gnu/share/config.site
touch /home/user/zcash/depends/i686-pc-linux-gnu/share/config.site
make: Leaving directory `/home/user/zcash/depends'
+ ./autogen.sh
libtoolize: putting auxiliary files in AC_CONFIG_AUX_DIR, `build-aux'.
libtoolize: copying file `build-aux/ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIR, `build-aux/m4'.
libtoolize: copying file `build-aux/m4/libtool.m4'
libtoolize: copying file `build-aux/m4/ltoptions.m4'
libtoolize: copying file `build-aux/m4/ltsugar.m4'
libtoolize: copying file `build-aux/m4/ltversion.m4'
libtoolize: copying file `build-aux/m4/lt~obsolete.m4'
libtoolize: putting auxiliary files in AC_CONFIG_AUX_DIR, `build-aux'.
libtoolize: copying file `build-aux/ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIR, `build-aux/m4'.
libtoolize: copying file `build-aux/m4/libtool.m4'
libtoolize: copying file `build-aux/m4/ltoptions.m4'
libtoolize: copying file `build-aux/m4/ltsugar.m4'
libtoolize: copying file `build-aux/m4/ltversion.m4'
libtoolize: copying file `build-aux/m4/lt~obsolete.m4'
Makefile.am:5: warning: user variable 'GZIP_ENV' defined here ...
/usr/share/automake-1.14/am/distdir.am: ... overrides Automake variable 'GZIP_ENV' defined here
Makefile.am:58: warning: user target 'distcleancheck' defined here ...
/usr/share/automake-1.14/am/distdir.am: ... overrides Automake target 'distcleancheck' defined here
src/Makefile.am:468: warning: user target '.mm.o' defined here ...
/usr/share/automake-1.14/am/depend2.am: ... overrides Automake target '.mm.o' defined here
+ ./configure --prefix=/home/user/zcash/depends/x86_64-unknown-linux-gnu/ --with-gui=no --enable-hardening '' 'CXXFLAGS=-Wno-deprecated-declarations -Wno-placement-new -Wno-terminate -Werror -O1 -g'
configure: WARNING: you should use --build, --host, --target
checking build system type... i686-pc-linux-gnu
checking host system type... i686-pc-linux-gnu
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... no
checking for mawk... mawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether to enable maintainer-specific portions of Makefiles... yes
checking whether make supports nested variables... (cached) yes
checking for g++... g++
checking whether the C++ compiler works... yes
checking for C++ compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking for style of include used by make... GNU
checking dependency style of g++... gcc3
checking for g++... g++
checking whether we are using the GNU Objective C++ compiler... no
checking whether g++ accepts -g... no
checking dependency style of g++... gcc3
checking how to print strings... printf
checking for gcc... gcc
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking dependency style of gcc... gcc3
checking for a sed that does not truncate output... /bin/sed
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for fgrep... /bin/grep -F
checking for ld used by gcc... /usr/bin/ld
checking if the linker (/usr/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 1572864
checking whether the shell understands some XSI constructs... yes
checking whether the shell understands "+="... yes
checking how to convert i686-pc-linux-gnu file names to i686-pc-linux-gnu format... func_convert_file_noop
checking how to convert i686-pc-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /usr/bin/ld option to reload object files... -r
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ar... ar
checking for archiver @FILE support... @
checking for strip... strip
checking for ranlib... ranlib
checking command to parse /usr/bin/nm -B output from gcc object... ok
checking for sysroot... no
checking for mt... mt
checking if mt is a manifest tool... no
checking how to run the C preprocessor... gcc -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... yes
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/usr/bin/ld) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking how to run the C++ preprocessor... g++ -E
checking for ld used by g++... /usr/bin/ld
checking if the linker (/usr/bin/ld) is GNU ld... yes
checking whether the g++ linker (/usr/bin/ld) supports shared libraries... yes
checking for g++ option to produce PIC... -fPIC -DPIC
checking if g++ PIC flag -fPIC -DPIC works... yes
checking if g++ static flag -static works... yes
checking if g++ supports -c -o file.o... yes
checking if g++ supports -c -o file.o... (cached) yes
checking whether the g++ linker (/usr/bin/ld) supports shared libraries... yes
checking dynamic linker characteristics... (cached) GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking for ar... /usr/bin/ar
checking for ranlib... /usr/bin/ranlib
checking for strip... /usr/bin/strip
checking for gcov... /usr/bin/gcov
checking for lcov... no
checking for java... no
checking for genhtml... no
checking for git... /usr/bin/git
checking for ccache... no
checking for xgettext... /usr/bin/xgettext
checking for hexdump... /usr/bin/hexdump
checking for readelf... /usr/bin/readelf
checking for c++filt... /usr/bin/c++filt
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
checking whether byte ordering is bigendian... no
checking if compiler needs -Werror to reject unknown flags... no
checking for the pthreads library -lpthreads... no
checking whether pthreads work without any flags... no
checking whether pthreads work with -Kthread... no
checking whether pthreads work with -kthread... no
checking for the pthreads library -llthread... no
checking whether pthreads work with -pthread... yes
checking for joinable pthread attribute... PTHREAD_CREATE_JOINABLE
checking if more special flags are required for pthreads... no
checking for PTHREAD_PRIO_INHERIT... yes
checking for special C compiler options needed for large files... no
checking for _FILE_OFFSET_BITS value needed for large files... 64
checking whether strerror_r is declared... yes
checking for strerror_r... yes
checking whether strerror_r returns char *... yes
checking whether the linker accepts -Wl,--large-address-aware... no
checking for __attribute__((visibility))... yes
checking for __attribute__((dllexport))... no
checking for __attribute__((dllimport))... no
checking whether C++ compiler accepts -Wstack-protector... yes
checking whether C++ compiler accepts -fstack-protector-all... yes
checking whether C++ preprocessor accepts -D_FORTIFY_SOURCE=2... yes
checking whether C++ preprocessor accepts -U_FORTIFY_SOURCE... yes
checking whether the linker accepts -Wl,-z,relro... yes
checking whether the linker accepts -Wl,-z,now... yes
checking whether C++ compiler accepts -fPIE... yes
checking whether the linker accepts -pie... yes
checking endian.h usability... yes
checking endian.h presence... yes
checking for endian.h... yes
checking sys/endian.h usability... no
checking sys/endian.h presence... no
checking for sys/endian.h... no
checking byteswap.h usability... yes
checking byteswap.h presence... yes
checking for byteswap.h... yes
checking stdio.h usability... yes
checking stdio.h presence... yes
checking for stdio.h... yes
checking for stdlib.h... (cached) yes
checking for unistd.h... (cached) yes
checking for strings.h... (cached) yes
checking for sys/types.h... (cached) yes
checking for sys/stat.h... (cached) yes
checking sys/select.h usability... yes
checking sys/select.h presence... yes
checking for sys/select.h... yes
checking sys/prctl.h usability... yes
checking sys/prctl.h presence... yes
checking for sys/prctl.h... yes
checking for library containing getaddrinfo_a... -lanl
checking for library containing inet_pton... none required
checking whether strnlen is declared... yes
checking whether le16toh is declared... yes
checking whether le32toh is declared... yes
checking whether le64toh is declared... yes
checking whether htole16 is declared... yes
checking whether htole32 is declared... yes
checking whether htole64 is declared... yes
checking whether be16toh is declared... yes
checking whether be32toh is declared... yes
checking whether be64toh is declared... yes
checking whether htobe16 is declared... yes
checking whether htobe32 is declared... yes
checking whether htobe64 is declared... yes
checking whether bswap_16 is declared... yes
checking whether bswap_32 is declared... yes
checking whether bswap_64 is declared... yes
checking for MSG_NOSIGNAL... yes
checking for library containing clock_gettime... none required
checking for visibility attribute... yes
checking for Berkeley DB C++ headers... default
checking for main in -ldb_cxx-4.8... yes
checking miniupnpc/miniwget.h usability... no
checking miniupnpc/miniwget.h presence... no
checking for miniupnpc/miniwget.h... no
checking miniupnpc/miniupnpc.h usability... no
checking miniupnpc/miniupnpc.h presence... no
checking for miniupnpc/miniupnpc.h... no
checking miniupnpc/upnpcommands.h usability... no
checking miniupnpc/upnpcommands.h presence... no
checking for miniupnpc/upnpcommands.h... no
checking miniupnpc/upnperrors.h usability... no
checking miniupnpc/upnperrors.h presence... no
checking for miniupnpc/upnperrors.h... no
checking whether to build Bitcoin Core GUI... no (Qt)
checking for boostlib >= 1.20.0... yes
checking whether the Boost::System library is available... yes
checking for exit in -lboost_system... yes
checking whether the Boost::Filesystem library is available... yes
checking for exit in -lboost_filesystem... yes
checking whether the Boost::Program_Options library is available... yes
checking for exit in -lboost_program_options... yes
checking whether the Boost::Thread library is available... yes
checking for exit in -lboost_thread... yes
checking whether the Boost::Chrono library is available... yes
checking for exit in -lboost_chrono... yes
checking whether the Boost::Unit_Test_Framework library is available... yes
checking for dynamic linked boost test... yes
checking for SSL... yes
checking for CRYPTO... yes
checking gmp.h usability... yes
checking gmp.h presence... yes
checking for gmp.h... yes
checking for __gmpn_sub_n in -lgmp... yes
checking gmpxx.h usability... yes
checking gmpxx.h presence... yes
checking for gmpxx.h... yes
checking for main in -lgmpxx... yes
Hunting for libsnark include directory...
configure: error: libsnark include directory not found
user@user-computer-756223:~/zcash$ ./zcutil/build.sh
```

**Comment by zcashuser** [comment]
I used these instructions to get BerkeleyDB working: https://github.com/bitcoin/bitcoin/blob/master/doc/build-unix.md#dependency-build-instructions-ubuntu--debian

**Comment by zcashuser** [comment]
The check for the libsnark include directory is:
```
echo 'Hunting for libsnark include directory...'
[LIBSNARK_INCDIR="$(echo "$CPPFLAGS" | sed 's,^.*-I\([^ ]*/include\).*$,\1/libsnark,')"]
if test -d "$LIBSNARK_INCDIR"; then
   echo "Found libsnark include directory: $LIBSNARK_INCDIR"
```
My $CPPVALUE at the time is: `-DBOOST_SPIRIT_THREADSAFE -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS -D_FILE_OFFSET_BITS=64  -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2`

**Comment by daira** [comment]
radix42 wrote on #zcash-dev slack:
> [...] the sed match works if the libsnark path is the only thing in CPPFLAGS, or at the front or end of the list of paths, I can't recall which (one of them worked, one didn't, when I had to build with additional CPPFLAGS for ./configure, can't recall which)

**Comment by daira** [comment]
Will be closed by moving libsnark in-tree (#2252).

**Comment by nathan-at-least** [comment]
I just bumped this out of 1.0.9, because this isn't a 1.0.9 goal *per se*. However, #2252 may be a goal, and as @daira said it would address this issue. I've placed this in the user support project to ensure we follow up on it.

**Comment by nathan-at-least** [comment]
This is blocked on setting up a xubuntu CI worker.

**Comment by str4d** [comment]
Closed by moving libsnark in-tree (#2252).
